### PR TITLE
Add a new tautomer mol hash

### PR DESCRIPTION
**Pull request by greglandrum**
This adds a new hash for doing tautomer-insensitive comparisons that handles keto-enol and imine-enamine tautomerism.

The new hash, `HetAtomTautomerv2` is inspired by the original `HetAtomTautomer` hash, but has a slightly different form (and was, unfortunately, more complex to implement). The new hash look likes this:

```
In [4]: rdMolHash.MolHash(Chem.MolFromSmiles('CCC=O'),rdMolHash.HashFunction.HetAtomTautomerv2)
Out[4]: '[CH3]-[C]:[C]:[O]_3_0'
```

All Hs and bonds are explicit in the hash. All Hs are removed from "tautomeric regions", counted, and reported at the end. The bonds in the "tautomeric regions" are all aromatic.

It's a bit slower than the `HetAtomTautomer` hash, but it's not too bad. Here are timings for generating the hashes for 50K ChEMBL molecules:
- `HetAtomTautomer`: 2.92s
- `HetAtomTautomerv2`: 3.33s

Since there's a potential order dependence in the algorithm, I did some bulk regression testing for this: I ran 2.2 million ChEMBL32 molecules through it and randomly renumbered atoms and bonds 10 times for each molecule. There were mismatches with 16 molecules and all of these are traceable to weird SMILES generation edge cases (need to be reported and fixed), but not to the changes made by the hashing algorithm.

This hash should be treated as "unstable" and likely to change from one release to the next for the moment.

### CODE REVIEW

**Comment by greglandrum** [comment]
@d-b-w @cdvonbargen I'm tagging you on this as well since you were also involved in these discussions.

**Review by cdvonbargen — COMMENTED** [review]
Do we want to add a toggle into RegistrationHash.py to use this? I guess a better question would be, if we don't, are we waiting until September to do so?

I also just noticed a couple references to Schro jira tickets in RegistrationHash.py that should probably remove....

**Comment by greglandrum** [comment]
> Do we want to add a toggle into RegistrationHash.py to use this? I guess a better question would be, if we don't, are we waiting until September to do so?
> 
> 

That's a good point. Maybe it's an extra layer (actually two layers) that we can add? So we'd have the current tautomer hash layers *and* the new ones.

> 
> I also just noticed a couple references to Schro jira tickets in RegistrationHash.py that should probably remove....

Yep. That would make sense.

**Comment by cdvonbargen** [comment]
Oh, instead of adding layers (and hash schemes), I was thinking we'd just:
```
def GetMolLayers(original_molecule, 
                 data_field_names = None,
                 escape = None, 
                 cxflag = DEFAULT_CXFLAG, 
                 enable_tautomer_v2 = False):
```
and which would populate HashLayer with either v1 or v2 tautomer data, and then HashScheme and GetMolHash would stay exactly the same.

**Comment by greglandrum** [comment]
I'm fine with that too.
I will make that change in the PR

File: rdkit/Chem/RegistrationHash.py [hunk 1]
```diff
@@ -55,9 +55,11 @@ class HashLayer(enum.Enum):
     :cvar ESCAPE: arbitrary other information to be incorporated
     :cvar FORMULA: a simple molecular formula for the molecule
     :cvar NO_STEREO_SMILES: RDKit canonical SMILES with all stereo removed
-    :cvar NO_STEREO_TAUTOMER_HASH: the above tautomer hash lacking all stereo
     :cvar SGROUP_DATA: canonicalization of all SGroups data present
     :cvar TAUTOMER_HASH: SMILES-like representation for a generic tautomer form
+    :cvar NO_STEREO_TAUTOMER_HASH: the above tautomer hash lacking all stereo
+    :cvar TAUTOMER_V2_HASH: SMILES-like representation for a generic tautomer form, v2
+    :cvar NO_STEREO_TAUTOMER_V2_HASH: the v2 tautomer hash lacking all stereo
```
Comment (cdvonbargen) [inline rdkit/Chem/RegistrationHash.py:62]: these aren't there, right?

**Review by cdvonbargen — APPROVED** [review]
Awesome, LGTM -- thanks Greg!

**Review by ricrogz — APPROVED** [review]
LGTM!

**Pull request merged**